src/compiler/scala/tools/nsc/typechecker/ConstantFolder.scala

```diff
@@ -155,7 +155,7 @@ abstract class ConstantFolder {
     case nme.GE  => Constant(x.intValue >= y.intValue)
     case nme.ADD => Constant(x.intValue + y.intValue)
     case nme.SUB => Constant(x.intValue - y.intValue)
-    case nme.MUL => Constant(x.intValue * y.intValue)
+    case nme.MUL => Constant(Math.multiplyExact(x.intValue, y.intValue))
```

Comment:
This doesn't address Seb's objection, that it throws and uses the original tree instead of the overflowing result.